Job selector: break out of the outer loop

troughton opened the pull request:
Without this, jobs can get lost by being removed from the waiting list but then replaced with a job from a lower-priority list.

septag commented:
Thank you!

I'm also profiling the job system, and the performance is inferior because of the naive implementation for global state which is all simple spin locks (job selector + wait). 
So there will be some major internal changes to data and state access and maybe convert some parts into lockless implementations. any suggestions or PRs are welcomed.

troughton commented:
It's not _particularly_ useful for `sx`, but I reimplemented the SX job system in Swift since that's what the rest of my code uses and it wasn't a huge amount of code – I know code-sharing with the main SX repo would be ideal, but this way was better ergonomically for the rest of the codebase. In the process of doing so, I reworked the synchronisation/wait lists in the way you suggest – per-thread wait lists and a global list.

I also switched to using a generation-based handle system as suggested [here](http://danglingpointers.com/post/job-system/), allowed partial dispatches of jobs, and fixed a couple of bugs.

If you're curious on how I implemented it/want to transfer any of it back into the main codebase, the code's all up here: https://gist.github.com/troughton/973820809c18f45b779105a5eb45c401. It should  map pretty directly to `jobs.c` minus the changes I've made – an `sx_job_t` is a `JobHandle` and an `sx__job` is a `JobItem`. If you've got questions about any part of it just leave a comment on the Gist and I'll do my best to answer it.
